dali/kernels/imgproc/jpeg/chroma_subsample_gpu.cuh

```diff
@@ -0,0 +1,241 @@
+}
+
+template <typename T>
+__inline__ __device__ vec<2, T> rgb_to_cb_cr(vec<3, T> rgb) {
+  return {rgb_to_cb<T>(rgb), rgb_to_cr<T>(rgb)};
+}
+
+template <int N, typename T>
+__inline__ __device__ vec<N> avg4(vec<N, T> a, vec<N, T> b, vec<N, T> c, vec<N, T> d) {
+  IMPL_VEC_ELEMENTWISE((a[i] + b[i] + c[i] + d[i]) * 0.25f);
+}
+
+template <int N, typename T>
+__inline__ __device__ vec<N> avg2(vec<N, T> a, vec<N, T> b) {
+  IMPL_VEC_ELEMENTWISE((a[i] + b[i]) * 0.5f);
+}
+
+template <typename T>
+__inline__ __device__ vec<3, T> ycbcr_to_rgb(const vec<3, T> ycbcr) {
+    float y  = static_cast<float>(ycbcr.x);
+    float cb = static_cast<float>(ycbcr.y) - 128.0f;
+    float cr = static_cast<float>(ycbcr.z) - 128.0f;
+    vec<3, T> rgb;
+    rgb.x = ConvertSat<T>(y + 1.402f * cr);
+    rgb.y = ConvertSat<T>(y - 0.34413629f * cb - 0.71413629f * cr);
+    rgb.z = ConvertSat<T>(y + 1.772f * cb);
+    return rgb;
+}
+
+template <bool horz_subsample, bool vert_subsample, typename T>
+__inline__ __device__
+void rgb_to_ycbcr_chroma_subsample(ivec2 blk_offset, ivec2 offset,
+                                   const Surface2D<T>& luma,
+                                   const Surface2D<T>& cb,
+                                   const Surface2D<T>& cr,
+                                   const Surface2D<const uint8_t>& in) {
+  const auto sampler = make_sampler<DALI_INTERP_NN>(in);
+  int chroma_y = blk_offset.y;
+  int chroma_x = blk_offset.x;
+  int luma_y = chroma_y << vert_subsample;
+  int luma_x = chroma_x << horz_subsample;
+  int y = offset.y;
+  int x = offset.x;
+
+  vec<3, T> rgb[4];
+  sampler(rgb[0].v, ivec2(x, y), BorderClamp());
+
+  int pixel_stride = &luma(1, 0) - &luma(0, 0);
+  int row_stride = &luma(0, 1) - &luma(0, 0);
+
+  luma(luma_x, luma_y) = rgb_to_y<T>(rgb[0]);
+
+  vec<3, T> avg_rgb(rgb[0]);
+  if (horz_subsample && vert_subsample) {
+    sampler(rgb[1].v, ivec2(x + 1, y), BorderClamp());
+    sampler(rgb[2].v, ivec2(x, y + 1), BorderClamp());
+    sampler(rgb[3].v, ivec2(x + 1, y + 1), BorderClamp());
+    luma(luma_x + 1, luma_y) = rgb_to_y<T>(rgb[1]);
+    luma(luma_x, luma_y + 1) = rgb_to_y<T>(rgb[2]);
+    luma(luma_x + 1, luma_y + 1) = rgb_to_y<T>(rgb[3]);
+    avg_rgb = avg4(rgb[0], rgb[1], rgb[2], rgb[3]);
+  } else if (horz_subsample) {
+    sampler(rgb[1].v, ivec2(x + 1, y), BorderClamp());
+    luma(luma_x + 1, luma_y) = rgb_to_y<T>(rgb[1]);
+    avg_rgb = avg2(rgb[0], rgb[1]);
+  } else if (vert_subsample) {
+    sampler(rgb[1].v, ivec2(x, y + 1), BorderClamp());
+    luma(luma_x, luma_y + 1) = rgb_to_y<T>(rgb[1]);
+    avg_rgb = avg2(rgb[0], rgb[1]);
+  }
+
+  vec<2, T> cbcr = rgb_to_cb_cr<T>(avg_rgb);
+  cb(chroma_x, chroma_y) = cbcr.x;
+  cr(chroma_x, chroma_y) = cbcr.y;
+}
+
+template <int N, typename T>
+__inline__ __device__
+void write_vec(T* ptr, vec<N, T> v) {
+  #pragma unroll
+  for (int i = 0; i < N; i++)
+    *ptr++ = v[i];
```

Comment:
```suggestion
  for (int i = 0; i < N; i++)
    ptr[i] = v[i];
```
I think that 2 kinds of indexing are more confusing both for the human reader and compiler.